BUGFIX: Fusion parser fix multi line comment

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

This fixes a bug where the Fusion parser would not parse following c-style comments correctly:

```
/**
comment with multiple stars even
**/
```

This happed when the ending count of `*` was even. So ending a comment with `***/` worked previously.

Now we use the "correct" regex from Jeffrey E.F. Friedl's book "Mastering Regular Expressions" Page 272 "Unrolling C Comments"
We already use his regex for string matching and it is really fast due to the unrolled loop. Faster than using the lazy quantifier `~^/\*.*?\*/~s`.

I did a performance test with 1 million iteration on three different comment samples (with each sample having a dynamic part to clear possible caches):

| Unrolled (this pr) | Simple Lazy Quantifier  |
|--------|--------|
| 0.143725s | 0.160235s |
| 0.181047s | 0.203759s |
| 0.156254s | 0.170144s | 


Additionally the error message for comments starting with `/**` was improved. Previously $nextLine->char(1) would return `**` instead of just one `*` because wrongly implemented.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Fusion parser fix multi line comment

mhsdesign commented on Neos.Fusion/Classes/Core/ObjectTreeParser/Lexer.php:
```diff
@@ -44,12 +44,14 @@ class Lexer
           (?:
-            \*[^/]          # if after the '*' there is a '/' break, else continue
-            [^*]*           # until the special case '*' is encountered - unrolled loop following Jeffrey Friedl
+            [^/*]           # break if its the end: '/'
```
im not sure why `*` is excluded in the character set. It works, and all tests pass without the exclusion. And there is no performance dropoff.

mhsdesign replied:
Im nearly sure that `[^/*]` must be enough and that `*` can never be encountered here.

because before we consume all `*` so there cant be one and also as last operation in the unrolled loop we consume all `*`. So independent of which entry point is used, the next char cannot be a `*`.

kitsunet replied:
Do we have tests for various variants of fusion file includes with glob patterns?

mhsdesign replied:
yes but this is about multiline comments. And this regex will do but i was just eager to understand everypart and think there is redundancy but nothing critical.

edit;

yes i discussed this with @grebaldi and we agreed that the regex can be simplified

mhsdesign committed:
BUGFIX: Fusion parser error message for unclosed comments

mhsdesign committed:
TASK: `MULTILINE_COMMENT` simplify part of the regex

The character exclusion: `[^/*]` can be simplified to `[^/]`.

The star `*` can never be encountered here, because independent of the entry point we previously consume all stars (`\*+`).

dlubitz commented on Neos.Fusion/Classes/Core/ObjectTreeParser/ExceptionMessage/MessageLinePart.php:
```diff
@@ -37,10 +37,10 @@ public function linePrint(int $offset = 0): string
     {
-        if ($index < 0) {
-            return mb_substr($this->linePart, $index, 1);
+        if (mb_strlen($this->linePart) < abs($index)) {
+            return '';
```
I don't get this code here. From my POV it is a difference, if you use a `-4` or a `4` to a 4 char long string. So shouldn't be the handling also be different?
`'abcd'` => `-4` ==> `'a'`
`'abcd'` => `4` ==> `''`

Caused by the `0` index, which also counts for the positive indexes.

I assume your tests are running as `mb_substr` "fixes" this.

mhsdesign replied:
yes this code was initially supposed to be an easy access on the string for building the exception message. But i messed it up a little :D Nothing api or anything further to worry about :D Thanks for looking into it though :D

> Additionally the error message for comments starting with `/**` was improved. Previously $nextLine->char(1) would return `**` instead of just one `*` because wrongly implemented.

dlubitz replied:
Ok, if the index < 0 exeeds the string start, it will always return the first char instead of empty string:
```
mb_substr("abcd", -99, 1) == "a"
```

But index > 0 and index > lenght always returns ""
```
mb_substr("abcd", 99, 1) == ""
```

That why an additional check for index < 0 is needed.

mhsdesign replied:
Thanks for commenting our slack discussion here :D
Implemented with https://github.com/neos/neos-development-collection/pull/4882/commits/196e600a6d75cf4b958349be0254ec6d2d9fdc1d

mhsdesign committed:
TASK: Document MessageLinePart::char

markusguenther reviewed (approved):
I can approve that it runs and does what is in the description. As I am not the fusion parser expert, I don't know if it is the best solution.
